govcd/adminvdc.go

```diff
@@ -52,33 +55,273 @@ func (adminVdc *AdminVdc) Refresh() error {
 func (adminVdc *AdminVdc) Update() (AdminVdc, error) {
-	task, err := adminVdc.UpdateAsync()
+	apiVersion, err := adminVdc.client.maxSupportedVersion()
 	if err != nil {
 		return AdminVdc{}, err
 	}
 
-	err = task.WaitTaskCompletion()
+	realFunction, ok := vdcProducerByVersion["vdc"+apiVersionToVcdVersion[apiVersion]]
+	if !ok {
+		return AdminVdc{}, fmt.Errorf("no entity type found %s", "vdc"+apiVersion)
+	}
+	if realFunction.UpdateVdc == nil {
+		return AdminVdc{}, fmt.Errorf("function UpdateVdc is not defined for %s", "vdc"+apiVersion)
+	}
+
+	util.Logger.Printf("[DEBUG] Update call function for version %s", realFunction.SupportedVersion)
+
+	updatedAdminVdc, err := realFunction.UpdateVdc(adminVdc)
 	if err != nil {
 		return AdminVdc{}, err
 	}
+	return *updatedAdminVdc, err
+}
+
+type vdcProducer struct {
+	SupportedVersion string
+	CreateVdc        func(adminOrg *AdminOrg, vdcConfiguration *types.VdcConfiguration) (*Vdc, error)
+	CreateVdcAsync   func(adminOrg *AdminOrg, vdcConfiguration *types.VdcConfiguration) (Task, error)
+	UpdateVdc        func(adminVdc *AdminVdc) (*AdminVdc, error)
+	UpdateVdcAsync   func(adminVdc *AdminVdc) (Task, error)
+}
+
+var vdcCrudV90 = vdcProducer{
+	SupportedVersion: "29.0",
+	CreateVdc:        createVdc,
+	CreateVdcAsync:   createVdcAsync,
+	UpdateVdc:        updateVdc,
+	UpdateVdcAsync:   updateVdcAsync,
+}
+
+var vdcCrudV97 = vdcProducer{
+	SupportedVersion: "32.0",
+	CreateVdc:        createVdcV97,
+	CreateVdcAsync:   createVdcAsyncV97,
+	UpdateVdc:        updateVdcV97,
+	UpdateVdcAsync:   updateVdcAsyncV97,
+}
+
+var vdcProducerByVersion = map[string]vdcProducer{
+	"vdc9.0":  vdcCrudV90,
+	"vdc9.1":  vdcCrudV90,
+	"vdc9.5":  vdcCrudV90,
+	"vdc9.7":  vdcCrudV97,
+	"vdc10.0": vdcCrudV97,
+}
+
+// CreateOrgVdc creates a VDC with the given params under the given organization
+// and waits for the asynchronous task to complete.
+// Returns an AdminVdc.
+func (adminOrg *AdminOrg) CreateOrgVdc(vdcConfiguration *types.VdcConfiguration) (*Vdc, error) {
+	apiVersion, err := adminOrg.client.maxSupportedVersion()
+	if err != nil {
+		return nil, err
+	}
+	realFunction, ok := vdcProducerByVersion["vdc"+apiVersionToVcdVersion[apiVersion]]
+	if !ok {
+		return nil, fmt.Errorf("no entity type found %s", "vdc"+apiVersion)
+	}
+	if realFunction.CreateVdc == nil {
+		return nil, fmt.Errorf("function CreateVdc is not defined for %s", "vdc"+apiVersion)
+	}
+
+	util.Logger.Printf("[DEBUG] CreateOrgVdc call function for version %s", realFunction.SupportedVersion)
+	return realFunction.CreateVdc(adminOrg, vdcConfiguration)
+}
+
+// CreateOrgVdcAsync creates a VDC with the given params under the given organization.
+// Returns an Task.
+func (adminOrg *AdminOrg) CreateOrgVdcAsync(vdcConfiguration *types.VdcConfiguration) (Task, error) {
+	apiVersion, err := adminOrg.client.maxSupportedVersion()
+	if err != nil {
+		return Task{}, err
+	}
+	realFunction, ok := vdcProducerByVersion["vdc"+apiVersionToVcdVersion[apiVersion]]
```

Comment:
Rather than realFunction, a better name here would be producer

Reply:
changed